module/zfs/dmu_objset.c

```diff
@@ -1072,6 +1079,58 @@ dmu_objset_snapshot_one(const char *fsname, const char *snapname)
 	return (err);
 }
 
+typedef struct dmu_objset_upgrade_args {
+	objset_t *dua_os;
+	int (*dua_cb)(objset_t *);
+} dmu_objset_upgrade_arg_t;
+
+static void
+dmu_objset_upgrade_trampoline(void *data)
+{
+	dmu_objset_upgrade_arg_t *arg = data;
+	objset_t *os = arg->dua_os;
+
+	cv_signal(&os->os_upgrade_cv);
+
+	mutex_enter(&os->os_upgrade_lock);
+	if (!os->os_upgrade_exit) {
+		mutex_exit(&os->os_upgrade_lock);
+
+		(void) arg->dua_cb(os);
+		mutex_enter(&os->os_upgrade_lock);
+	}
+	os->os_upgrade_thread = NULL;
+	cv_broadcast(&os->os_upgrade_cv);
+	mutex_exit(&os->os_upgrade_lock);
```

Comment:
> I don't see what's protecting the objset from being destroyed after that first cv_signal().

You can take a look at the implementation of ```dmu_objset_upgrade_stop()``` where it has to wait on the ```os_upgrade_cv``` and then take ```os_upgrade_lock```, if it's able to grab the lock, it means the upgrade thread must have released it. Plus the spinlock trick in the implementation of ```mutex_exit()```, I don't see any problems. Please let me know if I missed something.

> What's the alternative?

I would rather revive the interface to upgrade a dataset from CLI. If automatic upgrade to a dataset has ever run into problem, admins can reissue the command after they believe the problem has been fixed.